File: src/amusement/engine/orapa_mine/borders.py
"""Correspondance entre les libellés du livret (nombres 1-18, lettres
A-R) et les points d'entrée du plateau.

⚠️ Répartition provisoire, non confirmée visuellement avec certitude
(voir docs/plan.md). 18+18 = 36 correspond exactement au pourtour d'une
grille 9x9 (9+9 cases en haut/bas, 9+9 en gauche/droite), ce qui est
l'hypothèse retenue par défaut dans `board.BoardDimensions`.

Cette classe est volontairement isolée du moteur de rayon
(`raycast.fire_ray`), qui ne travaille que sur des positions/directions
brutes : si la vraie répartition des libellés s'avère différente, seule
cette classe a besoin d'être corrigée.
"""

from __future__ import annotations

from dataclasses import dataclass

from .board import BoardDimensions, Position
from .raycast import Direction


@dataclass(frozen=True)
class Entry:
    """Un point d'entrée : la case juste à l'extérieur du plateau d'où un
    rayon est tiré, et la direction dans laquelle il entre."""

    position: Position
    direction: Direction


class LabelScheme:
    """Traduit les libellés du livret ("15", "I", ...) en `Entry`."""

    def __init__(self, dimensions: BoardDimensions | None = None) -> None:
        dims = dimensions or BoardDimensions()
        self._label_to_entry: dict[str, Entry] = {}

        width, height = dims.width, dims.height

        # Bord haut : nombres 1..width, de gauche à droite, tir vers le bas.
        for col in range(width):
            self._label_to_entry[str(col + 1)] = Entry((col, -1), Direction.DOWN)
        # Bord bas : nombres width+1..2*width, de gauche à droite, tir vers le haut.
        for col in range(width):
            self._label_to_entry[str(width + col + 1)] = Entry((col, height), Direction.UP)
        # Bord gauche : lettres A.., de haut en bas, tir vers la droite.
        for row in range(height):
            self._label_to_entry[_letter(row)] = Entry((-1, row), Direction.RIGHT)
        # Bord droit : lettres suivantes, de haut en bas, tir vers la gauche.
        for row in range(height):
            self._label_to_entry[_letter(height + row)] = Entry((width, row), Direction.LEFT)

        self._entry_to_label = {v: k for k, v in self._label_to_entry.items()}

    def entry_for_label(self, label: str) -> Entry:
        try:
            return self._label_to_entry[label]
        except KeyError:
            raise ValueError(f"Libellé de point d'entrée inconnu : {label!r}") from None

    def label_for_entry(self, entry: Entry) -> str:
        try:
            return self._entry_to_label[entry]
        except KeyError:
            raise ValueError(f"Aucun libellé ne correspond à {entry!r}") from None

    def all_labels(self) -> list[str]:
        return list(self._label_to_entry)


def _letter(index: int) -> str:
    """0 -> 'A', 25 -> 'Z'. Suffisant pour des plateaux de taille raisonnable."""
    if not 0 <= index < 26:
        raise ValueError("Plateau trop grand pour un étiquetage A-Z simple.")
    return chr(ord("A") + index)

File: frontend/src/games/orapa-mine/borders.ts
/**
 * Port TypeScript de `borders.py` (mêmes hypothèses, voir docs/plan.md) :
 * traduit les libellés du livret (nombres/lettres) en points d'entrée
 * pour l'aperçu local (Phase 3). À garder synchronisé avec `borders.py`
 * si l'hypothèse de répartition change.
 */

import { type BoardDimensions, DEFAULT_DIMENSIONS, type Direction, type Position } from "./types";

export interface Entry {
  position: Position;
  direction: Direction;
}

export class LabelScheme {
  private labelToEntry = new Map<string, Entry>();
  private entryToLabel = new Map<string, string>();

  constructor(dimensions: BoardDimensions = DEFAULT_DIMENSIONS) {
    const { width, height } = dimensions;

    for (let col = 0; col < width; col++) {
      this.add(String(col + 1), { position: [col, -1], direction: "DOWN" });
    }
    for (let col = 0; col < width; col++) {
      this.add(String(width + col + 1), { position: [col, height], direction: "UP" });
    }
    for (let row = 0; row < height; row++) {
      this.add(letter(row), { position: [-1, row], direction: "RIGHT" });
    }
    for (let row = 0; row < height; row++) {
      this.add(letter(height + row), { position: [width, row], direction: "LEFT" });
    }
  }

  private add(label: string, entry: Entry): void {
    this.labelToEntry.set(label, entry);
    this.entryToLabel.set(entryKey(entry), label);
  }

  entryForLabel(label: string): Entry {
    const entry = this.labelToEntry.get(label);
    if (!entry) throw new Error(`Libellé de point d'entrée inconnu : ${label}`);
    return entry;
  }

  labelForEntry(entry: Entry): string {
    const label = this.entryToLabel.get(entryKey(entry));
    if (!label) throw new Error(`Aucun libellé ne correspond à ${JSON.stringify(entry)}`);
    return label;
  }

  allEntries(): Array<{ label: string; entry: Entry }> {
    return [...this.labelToEntry.entries()].map(([label, entry]) => ({ label, entry }));
  }
}

function entryKey(entry: Entry): string {
  return `${entry.position[0]},${entry.position[1]},${entry.direction}`;
}

function letter(index: number): string {
  if (index < 0 || index >= 26) throw new Error("Plateau trop grand pour un étiquetage A-Z simple.");
  return String.fromCharCode("A".charCodeAt(0) + index);
}
